app/src/main/java/org/mozilla/focus/webview/SystemWebView.kt

```diff
@@ -0,0 +1,330 @@
+        val connection = super.onCreateInputConnection(outAttrs)
+        outAttrs.imeOptions = outAttrs.imeOptions or ViewUtils.IME_FLAG_NO_PERSONALIZED_LEARNING
+        return connection
+    }
+
+    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
+        WebViewProvider.applyAppSettings(context, settings, this)
+    }
+
+    override fun onPause() {
+        super.onPause()
+        pauseTimers()
+    }
+
+    override fun onResume() {
+        super.onResume()
+        resumeTimers()
+    }
+
+    override fun restoreWebViewState(session: Session) {
+        val stateData = session.webViewState
+
+        val backForwardList: WebBackForwardList? = if (stateData != null)
+            super.restoreState(stateData)
+        else
+            null
+
+        val desiredURL = session.url.value
+
+        client.restoreState(stateData)
+        client.notifyCurrentURL(desiredURL)
+
+        // Pages are only added to the back/forward list when loading finishes. If a new page is
+        // loading when the Activity is paused/killed, then that page won't be in the list,
+        // and needs to be restored separately to the history list. We detect this by checking
+        // whether the last fully loaded page (getCurrentItem()) matches the last page that the
+        // WebView was actively loading (which was retrieved during onSaveInstanceState():
+        // WebView.getUrl() always returns the currently loading or loaded page).
+        // If the app is paused/killed before the initial page finished loading, then the entire
+        // list will be null - so we need to additionally check whether the list even exists.
+
+        if (backForwardList != null && backForwardList.currentItem!!.url == desiredURL) {
+            // restoreState doesn't actually load the current page, it just restores navigation history,
+            // so we also need to explicitly reload in this case:
+            reload()
+        } else {
+            loadUrl(desiredURL)
+        }
+    }
+
+    override fun saveWebViewState(session: Session) {
+        // We store the actual state into another bundle that we will keep in memory as long as this
+        // browsing session is active. The data that WebView stores in this bundle is too large for
+        // Android to save and restore as part of the state bundle.
+        val stateData = Bundle()
+
+        super.saveState(stateData)
+        client.saveState(this, stateData)
+
+        session.saveWebViewState(stateData)
+    }
+
+    override fun setBlockingEnabled(enabled: Boolean) {
+        client.isBlockingEnabled = enabled
+        if (enabled) {
+            WebViewProvider.applyAppSettings(context, settings, this)
+        } else {
+            WebViewProvider.disableBlocking(settings, this)
+        }
+
+        if (callback != null) {
+            callback!!.onBlockingStateChanged(enabled)
```

Comment:
Should we just do a safe call? `callback?.onBlock..`

Reply:
So this is something we'll see a lot, no doubt. Because callback is a var, Kotlin wisely surmises that code it doesn't know about might change its value to null from another thread. The simplest solution is to use callback?.let {} so the value inside that let block is no longer mutable.